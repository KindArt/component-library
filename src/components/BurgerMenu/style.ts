import styled, { css } from 'styled-components';
import { Colours } from '../../constants/styles';

interface IWrapper {
  isOpen?: boolean;
}

export const BurgerMenuWrapper = styled.div<IWrapper>`
  height: 20px;
  width: 20px;
  position: relative;
  color: ${Colours.TEXT_SECONDARY};

  ${({ isOpen }) => isOpen && css``}
`;

interface ILongLineSVG {
  isOpen: boolean;
}

export const TopLine = styled.svg<ILongLineSVG>`
  position: absolute;
  fill: currentColor;
  transform: rotate(0);
  transition: 0.2s top 0.4s, 0.1s left 0.2s, 0.2s transform, 0.2s opacity 0.2s;

  left: 0;
  top: 2px;

  ${({ isOpen }) =>
    isOpen &&
    css`
      transition: 0.2s opacity, 0.2s top, 0.2s left 0.2s, 0.2s transform 0.4s;

      right: auto;
      left: 3px;
      top: 9px;

      transform: rotate(-45deg);
    `}
`;

export const MiddleLine = styled.svg<ILongLineSVG>`
  position: absolute;
  fill: currentColor;
  transform: rotate(0);
  transition: 0.2s top 0.4s, 0.1s left 0.2s, 0.2s transform, 0.2s opacity 0.2s;

  top: 9px;
  left: 0;

  ${({ isOpen }) =>
    isOpen &&
    css`
      transition: 0.2s opacity, 0.2s top, 0.2s left 0.2s, 0.2s transform 0.4s;

      opacity: 0;
    `}
`;

export const BottomLine = styled.svg<ILongLineSVG>`
  position: absolute;
  fill: currentColor;
  transform: rotate(0);
  transition: 0.2s top 0.4s, 0.1s left 0.2s, 0.2s transform, 0.2s opacity 0.2s;

  left: 0;
  top: 16px;

  ${({ isOpen }) =>
    isOpen &&
    css`
      transition: 0.2s opacity, 0.2s top, 0.2s left 0.2s, 0.2s transform 0.4s;
      right: auto;
      left: 3px;
      top: 9px;

      transform: rotate(45deg);
    `}
`;
