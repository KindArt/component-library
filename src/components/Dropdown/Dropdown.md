Single Selection:

```tsx padded
const options = [
  { label: 'Label 1', value: 'Value 1' },
  { label: 'Label 2', value: 'Value 2' },
  { label: 'Label 3', value: 'Value 3' },
];
<Dropdown label="Label" placeholder="Placeholder" options={options} />;
```

Multiple Selection:

```tsx
const options = [
  { label: 'Label 1', value: 'Value 1' },
  { label: 'Label 2', value: 'Value 2' },
  { label: 'Label 3', value: 'Value 3' },
];
<Dropdown multipleSelection options={options} />;
```

Error state:

```jsx
const options = [
  { label: 'Label 1', value: 'Value 1' },
  { label: 'Label 2', value: 'Value 2' },
  { label: 'Label 3', value: 'Value 3' },
];
<Dropdown label="Label" errorMessage="There is an error" options={options} />;
```

Disabled:

```tsx
const options = [
  { label: 'Label 1', value: 'Value 1' },
  { label: 'Label 2', value: 'Value 2' },
  { label: 'Label 3', value: 'Value 3' },
];
<Dropdown label="Label" options={options} disabled />;
```
