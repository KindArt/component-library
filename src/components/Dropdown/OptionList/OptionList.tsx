import React, { FC } from 'react';
import { Button } from '../../';
import { DropdownOptionType, OptionValueType } from '../';
import {
  Label,
  Check,
  Wrapper,
  ButtonWrapperOverride,
  ContentWrapperOverride,
  OptionWrapper,
  List,
  Option,
} from './style';

interface IOptionListProps {
  selected: Array<OptionValueType>;
  options: Array<DropdownOptionType>;
  excluded?: Array<OptionValueType>;
  optionClick: (value: OptionValueType) => void;
}

const OptionList: FC<IOptionListProps> = ({ options, excluded, selected, optionClick }) => {
  const renderOptions = () => {
    const filteredOptions = options.filter((option) => {
      if (excluded && excluded.indexOf(option.value) > -1) {
        return false;
      }
      return true;
    });

    return filteredOptions.map((i) => {
      const indexIsSelected = selected.find((k) => i.value === k);

      return (
        <Option key={i.value} onClick={() => optionClick(i.value)}>
          <Button
            fullWidth
            classOverrides={{
              buttonWrapper: `${ButtonWrapperOverride}`,
              contentWrapper: `${ContentWrapperOverride}`,
            }}
          >
            <OptionWrapper>
              <Label>{i.label}</Label>
              {indexIsSelected ? <Check /> : null}
            </OptionWrapper>
          </Button>
        </Option>
      );
    });
  };

  return (
    <Wrapper>
      <List>{renderOptions()}</List>
    </Wrapper>
  );
};

export default OptionList;
