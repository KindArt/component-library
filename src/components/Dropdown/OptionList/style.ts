import styled, { css } from 'styled-components';
import { Colours, Typography } from '../../../constants/styles';

export const Wrapper = styled.div`
  box-shadow: 0 1px 2px 0 rgba(0, 0, 0, 0.08);
  border: 1px solid ${Colours.BORDER_COLOUR_DARKER};
  background: white;
  border-radius: 4px;
  transform: translateY(6px);
  max-height: 300px;
`;

export const List = styled.div`
  display: flex;
  flex-direction: column;
  max-height: 235px;
  position: relative;
  overflow-y: auto;
  -webkit-overflow-scrolling: touch;
`;

// needs the :focus
export const ButtonWrapperOverride = css`
  outline: none;
  cursor: pointer;
  background: transparent;
  border: none;
  border-radius: 0;
`;

export const ContentWrapperOverride = css`
  flex-grow: 0;
`;

export const OptionWrapper = styled.div`
  display: flex;
  align-items: center;
  & > * {
    min-width: 1px;
    text-overflow: ellipsis;
    white-space: nowrap;
  }
`;

export const Label = styled.div`
  font-size: ${Typography.DEFAULT_FONT_SIZE};
  pointer-events: none;
`;

export const Check = styled.div`
  color: ${Colours.GREEN};
  margin-left: 12px;
  flex-shrink: 0;
  position: relative;
  width: 5px;
  height: 5px;
  background: ${Colours.GREEN};
  border-radius: 50%;
`;
