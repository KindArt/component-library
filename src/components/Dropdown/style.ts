import styled, { css } from 'styled-components';
import { Colours, CssVariables } from '../../constants/styles';

// Dropdown Styles

export const Wrapper = styled.div`
  ${({ className }) =>
    className &&
    css`
      ${className}
    `}
`;

export const ValueRendered = css`
  width: 100%;
  text-align: left; // important for other selectors - :hover etc.
  position: relative;
  background: #ffffff;
  border: ${CssVariables.BORDER_DEFAULT};
  box-shadow: 0 1px 2px 0 rgba(0, 0, 0, 0.08);
`;

export const ValueRenderedWrapper = styled.div`
  display: flex;
  align-items: center;
  justify-content: center;
  width: 100%;
  & > * {
    flex-grow: 1;
    max-width: 100%;
    min-width: 1px;
  }
`;

export const ArrowWrapper = styled.div`
  flex-grow: 0;
  transition: 0.2s color;
  position: relative;
  width: 1px;
  height: 100%;
  margin-left: 12px;
  &:before,
  &:after {
    display: block;
    position: absolute;
    content: '';
    top: 0;
    left: 0;
    width: 0;
    height: 0;
    border-left: 3px solid transparent;
    border-right: 3px solid transparent;
    border-bottom: 3px solid ${Colours.TEXT_SECONDARY};
  }
  &:after {
    top: -4px;
  }
  &:before {
    top: 2px;
    transform: rotate(180deg);
  }
`;

const PlaceholderEmpty = css`
  color: ${Colours.TEXT_SECONDARY};
`;

interface IValueWrapper {
  placeholderEmpty?: boolean;
}

export const ValueWrapper = styled.div<IValueWrapper>`
  text-overflow: ellipsis;
  white-space: nowrap;
  ${({ placeholderEmpty }) => placeholderEmpty && PlaceholderEmpty}
`;

export const PlaceholderAndOther = styled.div`
  display: flex;
  align-items: center;
  justify-content: flex-start;
  & > *:first-child {
    min-width: 5px;
    text-overflow: ellipsis;
    white-space: nowrap;
  }
`;

export const PlusValue = styled.div`
  display: inline-block;
  background: ${Colours.BORDER_COLOUR_DARKER};
  border: ${CssVariables.BORDER_DEFAULT};
  height: 20px;
  border-radius: 100px;
  display: flex;
  justify-content: center;
  align-items: center;
  padding: 0 5px;
  min-width: 30px;
  margin-left: 10px;
  font-size: 0.7rem;
`;

export const ErrorMessage = styled.p`
  color: ${({ theme }) => theme.errorMessage.color};
  margin-top: 6px;
  line-height: 1.375rem;
`;

ErrorMessage.defaultProps = {
  theme: {
    errorMessage: {
      color: `${Colours.ERROR}`,
    },
  },
};
