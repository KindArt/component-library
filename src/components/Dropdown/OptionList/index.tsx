import React, { FC } from 'react';
import { Button } from '../../';
import { DropdownOptionType, DropdownOptionValueType } from '../';
import { Label, Check, Wrapper, ButtonWrapperOverride, ContentWrapperOverride, OptionWrapper, List } from './style';

interface IOptionListProps {
  selected: Array<DropdownOptionType>;
  options: Array<DropdownOptionType>;
  excluded?: Array<DropdownOptionValueType>;
  optionClick: Function;
  optionProps?: object;
}

const OptionList: FC<IOptionListProps> = ({ options, excluded, selected, optionClick }) => {
  const renderOptions = () => {
    const filteredOptions = options.filter(option => {
      if (excluded && excluded.indexOf(option.value) > -1) {
        return false;
      }
      return true;
    });

    return filteredOptions.map(i => {
      const indexIsSelected = selected.find(k => i.value === k.value);
      const onClick = (key: DropdownOptionType) => () => {
        optionClick(key);
      };
      return (
        <Button
          key={i.value}
          classOverrides={{
            buttonWrapper: `${ButtonWrapperOverride}`,
            contentWrapper: `${ContentWrapperOverride}`
          }}
          onClick={onClick(i)}
        >
          <OptionWrapper>
            <Label>{i.label}</Label>

            {indexIsSelected ? <Check /> : null}
          </OptionWrapper>
        </Button>
      );
    });
  };

  return (
    <Wrapper>
      <List>{renderOptions()}</List>
    </Wrapper>
  );
};

export default OptionList;
