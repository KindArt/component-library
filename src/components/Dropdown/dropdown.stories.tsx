import React from 'react';
import Dropdown, { DropdownOptionType } from '.';

export default {
  component: Dropdown,
  title: 'Dropdown',
  decorators: [(storyFn: any) => <div style={{ maxWidth: '50%' }}>{storyFn()}</div>]
};

const optionsMock: Array<DropdownOptionType> = [
  { label: 'GBP', value: '1' },
  { label: 'CHF', value: '2' },
  { label: 'JPY', value: '3' }
];

export const normal = () => <Dropdown options={optionsMock} />;

export const multipleSelection = () => <Dropdown options={optionsMock} multipleSelection />;

export const empty = () => <Dropdown options={[]} />;

export const errorState = () => <Dropdown options={optionsMock} errorMessage="This is an error!" />;
