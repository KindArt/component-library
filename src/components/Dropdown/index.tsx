import React, { FC, HTMLProps, ReactNode, useState } from 'react';
import { Button, Label } from '../';
import OptionList from './OptionList/OptionList';
import OutsideClickHandler from '../../hocs/OutsideClickHandler';

import {
  Wrapper,
  ValueRendered,
  ValueRenderedWrapper,
  ArrowWrapper,
  ValueWrapper,
  PlaceholderAndOther,
  PlusValue,
  ErrorMessage,
} from './style';

export type DropdownOptionValueType = string | number;

export type DropdownOptionType = {
  label: ReactNode;
  value: DropdownOptionValueType;
};

// Custom value is used so omit it
export interface IDropdownProps extends Omit<HTMLProps<HTMLSelectElement>, 'value'> {
  options: Array<DropdownOptionType>;
  defaultOption?: Array<DropdownOptionValueType> | DropdownOptionValueType;
  value?: Array<DropdownOptionType>;
  excluded?: Array<DropdownOptionValueType>;
  multipleSelection?: boolean;
  optionProps?: object;
  label?: string;
  disabled?: boolean;
  errorMessage?: string;
  callbackOnChange?: Function;
  classOverrides?: {
    wrapper?: string;
  };
}

const Dropdown: FC<IDropdownProps> = ({
  defaultOption,
  options,
  label,
  disabled,
  placeholder,
  excluded,
  multipleSelection,
  optionProps,
  value,
  errorMessage,
  callbackOnChange,
  classOverrides,
}) => {
  const defaultValue = Array.isArray(defaultOption) ? defaultOption : [defaultOption];

  const selectedOptions: Array<DropdownOptionType> = [];
  options.forEach((item) => {
    if (defaultValue.indexOf(item.value) !== -1) {
      selectedOptions.push(item);
    }
  });

  const [selected, setSelected] = useState(selectedOptions);
  const [focused, setFocused] = useState(false);

  const onToggleFocus = () => setFocused(!focused);

  const renderValue = () => {
    const firstRender = selected[0];
    if (typeof firstRender !== 'undefined') {
      if (!firstRender.label) {
        const full = options.find((item) => firstRender.value === item.value);
        if (full) {
          firstRender.label = full.label;
        }
      }
    }

    if (options.length < 1) {
      return <ValueWrapper>No options available</ValueWrapper>;
    }

    if (disabled || selected.length < 1) {
      return <ValueWrapper placeholderEmpty={true}>{placeholder}</ValueWrapper>;
    }

    if (selected.length === 1) {
      return <ValueWrapper>{firstRender.label}</ValueWrapper>;
    }

    return (
      <PlaceholderAndOther>
        <div>{firstRender.label}</div>
        <PlusValue>+{selected.length - 1}</PlusValue>
      </PlaceholderAndOther>
    );
  };

  const showOptionList = focused && !disabled;

  // const handleOnClickAll = (clearOnly = false) => {
  //   if (selected >= options || clearOnly) {
  //     // this.callCallbackIfAvailable = true;
  //     setSelected([]);
  //   } else {
  //     const selected = options.map(o => o);
  //     // this.onChange()
  //     setSelected(selected);
  //   }
  // };

  const setSelectedStateAndCallback = (selected: Array<DropdownOptionType>) => {
    setSelected(selected);
    callbackOnChange && callbackOnChange(selected);
  };

  const handleOptionClick = (option: DropdownOptionType) => {
    if (!multipleSelection) {
      setSelectedStateAndCallback([option]);
      setFocused(false);
    } else {
      const selectedClone = [...selected];
      const index = selectedClone.map((item) => item.value).indexOf(option.value);
      if (index !== -1) {
        selectedClone.splice(index, 1);
      } else {
        selectedClone.push(option);
      }
      setSelectedStateAndCallback(selectedClone);
    }
  };

  return (
    <OutsideClickHandler onOutsideClick={focused ? () => setFocused(false) : null} onOutsideScroll={false}>
      <Wrapper className={classOverrides && classOverrides.wrapper}>
        {label && <Label content={label} />}
        <Button
          classOverrides={{
            buttonWrapper: `${ValueRendered}`,
          }}
          disabled={disabled || options.length < 1}
          onClick={onToggleFocus}
          error={!!errorMessage}
        >
          <ValueRenderedWrapper>
            {renderValue()}
            <ArrowWrapper />
          </ValueRenderedWrapper>
        </Button>
        {errorMessage && <ErrorMessage>{errorMessage}</ErrorMessage>}
        {showOptionList && (
          <OptionList
            excluded={excluded}
            optionClick={handleOptionClick}
            optionProps={optionProps}
            options={options}
            selected={value || selected}
          />
        )}
      </Wrapper>
    </OutsideClickHandler>
  );
};

export default Dropdown;
