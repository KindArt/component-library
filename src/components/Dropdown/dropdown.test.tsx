import React from 'react';
import { shallow } from 'enzyme';
import Dropdown, { IDropdownProps } from '.';
import { Label, Button } from '../';
import OptionList from './OptionList';
import { ErrorMessage } from './style';

const defaultProps: IDropdownProps = {
  options: [{ label: 'A label', value: '1' }]
};

describe('<Dropdown />', () => {
  let wrapper: any;
  beforeEach(() => {
    wrapper = shallow(<Dropdown {...defaultProps} />);
  });

  it('should render', () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe('Visuals', () => {
    it('should display a label if one is passed', () => {
      wrapper.setProps({ label: 'Dropdown' });
      const actual = wrapper.find(Label).length;
      expect(actual).toEqual(1);
    });

    it('should have the button component', () => {
      expect(wrapper.find(Button).length).toEqual(1);
    });

    it('should not display the option list by default', () => {
      expect(wrapper.find(OptionList).length).toEqual(0);
    });

    it('should display the option list when the button is clicked', () => {
      wrapper.find(Button).simulate('click');
      expect(wrapper.find(OptionList).length).toEqual(1);
    });

    it('should not display an error message by default', () => {
      expect(wrapper.find(ErrorMessage).length).toEqual(0);
    });

    it('should display an error message if one is passed', () => {
      wrapper.setProps({ errorMessage: 'This is an error.' });
      expect(wrapper.find(ErrorMessage).length).toEqual(1);
    });

    it('should not show the options list when clicked if disabled is true', () => {
      wrapper.setProps({ disabled: true });
      wrapper.find(Button).simulate('click');
      expect(wrapper.find(OptionList).length).toEqual(0);
    });
  });
});
