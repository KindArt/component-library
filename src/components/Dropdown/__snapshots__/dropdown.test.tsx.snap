// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Dropdown /> should render 1`] = `
<OutsideClickHandler
  Component="div"
  onOutsideClick={null}
  onOutsideScroll={false}
  onWindowResize={true}
>
  <styled.div>
    <Button
      classOverrides={
        Object {
          "buttonWrapper": "
  width: 100%;
  text-align: left; // important for other selectors - :hover etc.
  position: relative;
  background: #ffffff;
  border: 1px solid ,#D8DCE6,;
  box-shadow: 0 1px 2px 0 rgba(0, 0, 0, 0.08);
",
        }
      }
      disabled={false}
      error={false}
      onClick={[Function]}
    >
      <styled.div>
        <styled.div
          placeholderEmpty={true}
        />
        <styled.div />
      </styled.div>
    </Button>
  </styled.div>
</OutsideClickHandler>
`;
